src/mpi/errhan/errutil.c

```diff
@@ -1024,16 +1024,21 @@ int MPIR_Err_create_code_valist(int lastcode, int fatal, const char fcname[],
                                 const char generic_msg[], const char specific_msg[], va_list Argp)
 {
     int generic_idx;
-    int errcode = lastcode;
+    int err_code;
     if (lastcode == MPI_SUCCESS) {
         generic_idx = FindGenericMsgIndex(generic_msg);
+        err_code = error_class;
         if (generic_idx >= 0) {
-            errcode = (generic_idx << ERROR_GENERIC_SHIFT) | error_class;
-            if (fatal)
-                errcode |= ERROR_FATAL_MASK;
+            err_code += (generic_idx + 1) << ERROR_GENERIC_SHIFT;
         }
+        if (fatal)
+            err_code |= ERROR_FATAL_MASK;
+    } else if (is_valid_error_class(lastcode) || (lastcode & ERROR_DYN_MASK)) {
+        err_code = lastcode;
+    } else {
+        err_code = error_class;
```

Comment:
Could you explain this change? (maybe we should schedule a call to walk me through).

Reply:
The changes here are to mimic the changes in the other conditionally-compiled routine with the same name. We want the `lastcode` from user callbacks to pass through, but only if its value corresponds to a valid error class, otherwise jut return `error_class`.

Reply:
In other words, the last two branches are just an attempt to NOT pass through values that are unknown to MPI. To pass through, `lastcode` has to either b A) a predefined error class `MPI_ERR_*`, or B) a dynamically created error class or code. If a user callback returns let say `return 123456`, that thing will not pass through, and MPI will eventually report back the `error_class` that was passed to  `MPIR_Err_create_code_valist()`.